smart_contract/src/lib.rs

```diff
@@ -79,72 +78,70 @@ pub trait ExecuteOnHost: Instruction {
     fn execute(&self) -> Result<(), ValidationFail>;
 }
 
-// TODO: Remove the Clone bound. It can be done by custom serialization to InstructionExpr
-impl<I: Instruction + Encode + Clone> ExecuteOnHost for I {
+impl<I: Instruction + Encode> ExecuteOnHost for I {
     fn execute(&self) -> Result<(), ValidationFail> {
         #[cfg(not(test))]
         use host::execute_instruction as host_execute_instruction;
         #[cfg(test)]
         use tests::_iroha_smart_contract_execute_instruction_mock as host_execute_instruction;
 
-        // TODO: Redundant conversion into `InstructionExpr`
-        let isi_box: InstructionBox = self.clone().into();
+        let bytes = self.encode_as_instruction_box();
         // Safety: `host_execute_instruction` doesn't take ownership of it's pointer parameter
         unsafe {
-            decode_with_length_prefix_from_raw(encode_and_execute(
-                &isi_box,
-                host_execute_instruction,
+            decode_with_length_prefix_from_raw(host_execute_instruction(
+                bytes.as_ptr(),
+                bytes.len(),
             ))
         }
     }
 }
 
+#[derive(Debug, Encode)]
+enum QueryRequest<'a, Q> {
+    Query(QueryWithParameters<'a, Q>),
+    Cursor(&'a ForwardCursor),
+}
+
 /// Generic query request containing additional parameters.
 #[derive(Debug)]
-pub struct QueryRequest<Q> {
-    query: Q,
+pub struct QueryWithParameters<'a, Q> {
+    query: &'a Q,
     sorting: Sorting,
     pagination: Pagination,
     fetch_size: FetchSize,
 }
 
-impl<Q: Query> From<QueryRequest<Q>> for SmartContractQueryRequest {
-    fn from(query_request: QueryRequest<Q>) -> Self {
-        SmartContractQueryRequest::query(
-            query_request.query.into(),
-            query_request.sorting,
-            query_request.pagination,
-            query_request.fetch_size,
-        )
+impl<Q: Query> Encode for QueryWithParameters<'_, Q> {
```

Comment:
I think having common structure in data_model with implemented `EncodeLike` would be more robust.

Currently there's no state that this encoding should be fully compatible with the structure in `core/src/smartcontracts/wasm.rs`